x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/oidc/OpenIdConnectAuthenticator.java

```diff
@@ -0,0 +1,434 @@
+        } else if (state.equals(expectedState) == false) {
+            logger.debug("Invalid state parameter [{}], while [{}] was expected", state, expectedState);
+            throw new ElasticsearchSecurityException("Received a response with an invalid state parameter.");
+        }
+    }
+
+    /**
+     * Wraps sending an {@link HTTPRequest} with the appropriate permissions and returns the associated {@link HTTPResponse}
+     */
+    private HTTPResponse getResponse(HTTPRequest httpRequest) throws PrivilegedActionException {
+        SpecialPermission.check();
+        return AccessController.doPrivileged((PrivilegedExceptionAction<HTTPResponse>) httpRequest::send);
+    }
+
+    /**
+     * Converts an {@link AbstractRequest} to a {@link HTTPRequest} setting the necessary TLS configuration and timeout parameters
+     */
+    private HTTPRequest buildRequest(AbstractRequest request) {
+        final String sslKey = RealmSettings.realmSslPrefix(realmConfig.identifier());
+        final SSLConfiguration sslConfiguration = sslService.getSSLConfiguration(sslKey);
+        boolean isHostnameVerificationEnabled = sslConfiguration.verificationMode().isHostnameVerificationEnabled();
+        final HostnameVerifier verifier = isHostnameVerificationEnabled ? new DefaultHostnameVerifier() : NoopHostnameVerifier.INSTANCE;
+        final HTTPRequest httpRequest = request.toHTTPRequest();
+        httpRequest.setSSLSocketFactory(sslService.sslSocketFactory(sslConfiguration));
+        httpRequest.setHostnameVerifier(verifier);
+        httpRequest.setConnectTimeout(Math.toIntExact(realmConfig.getSetting(HTTP_CONNECT_TIMEOUT).getMillis()));
+        httpRequest.setReadTimeout(Math.toIntExact(realmConfig.getSetting(HTTP_CONNECTION_READ_TIMEOUT).getSeconds()));
+        return httpRequest;
+    }
+
+    /**
+     * Completes the Authorization Code Grant authentication flow of OpenId Connect by exchanging the received
+     * authorization code for an Id Token and an access token.
+     *
+     * @param code the authorization code that was received as a response to the authentication request
+     * @return a {@link Tuple} containing the received (yet not validated) {@link AccessToken} and {@link JWT}
+     */
+    private Tuple<AccessToken, JWT> exchangeCodeForToken(AuthorizationCode code) {
+        Secret clientSecret = null;
+        try {
+            clientSecret = new Secret(rpConfig.getClientSecret().toString());
+            final ClientAuthentication clientAuth = new ClientSecretBasic(rpConfig.getClientId(), clientSecret);
+            final AuthorizationGrant codeGrant = new AuthorizationCodeGrant(code, rpConfig.getRedirectUri());
+            final TokenRequest tokenRequest = new TokenRequest(opConfig.getTokenEndpoint(), clientAuth, codeGrant);
+            final HTTPResponse httpResponse = getResponse(buildRequest(tokenRequest));
```

Comment:
as @albertzaharovits mentioned already, we need to make this asynchronous. We have a few options:

1. fork to the generic threadpool and execute this blocking call there
2. add HTTPRequest#send as a forbidden api, add http-asyncclient as a dependency and use it to execute asynchronously.

I'm in favor of option 2 even though it is more work. It feels more like the right thing to do.

Reply:
I went with option 2

Reply:
Yeah I do think we should. The reason being that it might not be obvious to someone else that it should not be used